tensorflow/training/docker/2.5/py3/cu112/Dockerfile.gpu

```diff
@@ -0,0 +1,330 @@
+   && tar -xzf boost_1_73_0.tar.gz \
+   && cd boost_1_73_0 \
+   && ./bootstrap.sh \
+   && ./b2 threading=multi --prefix=/usr -j 64 cxxflags=-fPIC cflags=-fPIC install || true \
+   && cd .. \
+   && rm -rf boost_1_73_0.tar.gz \
+   && rm -rf boost_1_73_0 \
+   && cd /usr/include/boost
+
+###########################################################################
+# Horovod & its dependencies
+###########################################################################
+
+# Create a wrapper for OpenMPI to allow running as root by default
+RUN mv $OPEN_MPI_PATH/bin/mpirun $OPEN_MPI_PATH/bin/mpirun.real \
+ && echo '#!/bin/bash' > $OPEN_MPI_PATH/bin/mpirun \
+ && echo 'mpirun.real --allow-run-as-root "$@"' >> $OPEN_MPI_PATH/bin/mpirun \
+ && chmod a+x $OPEN_MPI_PATH/bin/mpirun
+
+# Configure OpenMPI to run good defaults:
+#   --bind-to none --map-by slot --mca btl_tcp_if_exclude lo,docker0
+RUN echo "hwloc_base_binding_policy = none" >> $OPEN_MPI_PATH/etc/openmpi-mca-params.conf \
+ && echo "rmaps_base_mapping_policy = slot" >> $OPEN_MPI_PATH/etc/openmpi-mca-params.conf
+
+# Set default NCCL parameters
+RUN echo NCCL_DEBUG=INFO >> /etc/nccl.conf
+ENV LD_LIBRARY_PATH=$OPEN_MPI_PATH/lib/:$EFA_PATH/lib/:$LD_LIBRARY_PATH
+# /usr/local/lib/libpython* needs to be accessible for dynamic linking
+ENV LD_LIBRARY_PATH=/usr/local/lib:$LD_LIBRARY_PATH
+ENV PATH=$OPEN_MPI_PATH/bin/:$PATH
+ENV PATH=$OPEN_MPI_PATH/nvidia/bin:$PATH
+
+# SSH login fix. Otherwise user is kicked off after login
+RUN mkdir -p /var/run/sshd \
+   && sed 's@session\s*required\s*pam_loginuid.so@session optional pam_loginuid.so@g' -i /etc/pam.d/sshd
+
+# Create SSH key.
+RUN mkdir -p /root/.ssh/ \
+   && ssh-keygen -q -t rsa -N '' -f /root/.ssh/id_rsa \
+   && cp /root/.ssh/id_rsa.pub /root/.ssh/authorized_keys \
+   && printf "Host *\n  StrictHostKeyChecking no\n" >> /root/.ssh/config
+
+WORKDIR /
+
+RUN apt-get update \
+   && apt-get install -y --no-install-recommends \
+   libbz2-dev \
+   libc6-dev \
+   libffi-dev \
+   libgdbm-dev \
+   libncursesw5-dev \
+   libreadline-gplv2-dev \
+   libsqlite3-dev \
+   libssl-dev \
+   tk-dev \
+   && rm -rf /var/lib/apt/lists/* \
+   && apt-get clean
+
+RUN wget https://www.python.org/ftp/python/$PYTHON_VERSION/Python-$PYTHON_VERSION.tgz \
+   && tar -xvf Python-$PYTHON_VERSION.tgz \
+   && cd Python-$PYTHON_VERSION \
+   && ./configure --enable-shared \
+   && make && make install \
+   && rm -rf ../Python-$PYTHON_VERSION*
+
+RUN ${PIP} --no-cache-dir install --upgrade \
+   pip \
+   setuptools
+
+RUN wget -c https://www.openssl.org/source/openssl-${OPENSSL_VERSION}.tar.gz \
+   && apt-get update \
+   && apt remove -y --purge openssl \
+   && rm -rf /usr/include/openssl \
+   && apt-get install -y \
+   ca-certificates \
+   && tar -xzvf openssl-${OPENSSL_VERSION}.tar.gz \
+   && cd openssl-${OPENSSL_VERSION} \
+   && ./config && make && make test \
+   && make install \
+   && ldconfig \
+   && cd .. && rm -rf openssl-*
```

Comment:
Why are we building openssl 1.1.1g from source?

default ubuntu 18.04 cu112 nvidiadockerfile contains openssl 1.1.1 pre-installed
```
$ docker run -it nvidia/cuda:11.2.1-base-ubuntu18.04 bash
openssl version
OpenSSL 1.1.1  11 Sep 2018
```

Previous dockerfiles relied on default apt distribution of openssl.

Reply:
done